clients/pkg/promtail/client/watcher.go

```diff
@@ -0,0 +1,326 @@
+// is reading / tailing segments. Also, when a new segment is detected, and the watcher is moving on from one to the other,
+// the SegmentEnd callback is available to take action if required.
+// The implementor of this interface is not expected to implement thread safety if used on a single watcher, since each
+// callback is called synchronously.
+type WatcherConsumer interface {
+	ConsumeSeries(series record.RefSeries) error
+	ConsumeEntries(entries ingester.RefEntries) error
+	SegmentEnd(segmentNum int)
+}
+
+type WALWatcher struct {
+	consumer   WatcherConsumer
+	done       chan struct{}
+	quit       chan struct{}
+	walDir     string
+	logger     log.Logger
+	MaxSegment int
+}
+
+func NewWALWatcher(walDir string, consumer WatcherConsumer, logger log.Logger) *WALWatcher {
+	return &WALWatcher{
+		walDir:     walDir,
+		consumer:   consumer,
+		quit:       make(chan struct{}),
+		done:       make(chan struct{}),
+		MaxSegment: -1,
+		logger:     logger,
+	}
+}
+
+// Start runs the watcher main loop.
+func (w *WALWatcher) Start() {
+	go w.mainLoop()
+}
+
+func (w *WALWatcher) Stop() {
+	// first close the quit channel to order main mainLoop routine to stop
+	close(w.quit)
+	// upon calling stop, wait for main mainLoop execution to stop
+	<-w.done
+}
+
+func (w *WALWatcher) mainLoop() {
+	defer close(w.done)
+	for !isClosed(w.quit) {
+		//w.SetStartTime(time.Now())
+		if err := w.run(); err != nil {
+			level.Error(w.logger).Log("msg", "error tailing WAL", "err", err)
+		}
+
+		select {
+		case <-w.quit:
+			return
+		case <-time.After(5 * time.Second):
+		}
+	}
+}
+
+// Run the watcher, which will tail the WAL until the quit channel is closed
+// or an error case is hit.
+func (w *WALWatcher) run() error {
+	_, lastSegment, err := w.firstAndLast()
+	if err != nil {
+		return fmt.Errorf("wal.Segments: %w", err)
+	}
+
+	currentSegment := lastSegment
+	//level.Debug(w.logger).Log("msg", "Tailing WAL", "lastCheckpoint", lastCheckpoint, "checkpointIndex", checkpointIndex, "currentSegment", currentSegment, "lastSegment", lastSegment)
+	level.Debug(w.logger).Log("msg", "Tailing WAL", "currentSegment", currentSegment, "lastSegment", lastSegment)
+	for !isClosed(w.quit) {
+		//w.currentSegmentMetric.Set(float64(currentSegment))
+		level.Debug(w.logger).Log("msg", "Processing segment", "currentSegment", currentSegment)
+
+		// On start, after reading the existing WAL for series records, we have a pointer to what is the latest segment.
+		// On subsequent calls to this function, currentSegment will have been incremented and we should open that segment.
+		if err := w.watch(currentSegment, currentSegment >= lastSegment); err != nil {
+			return err
+		}
+
+		// For testing: stop when you hit a specific segment.
+		if currentSegment == w.MaxSegment {
+			return nil
+		}
+
+		// we now a new segment has been cut, upon advancing the segment pointer, emit the send batch call
```

Comment:
This is only true because we're doing a single batch per segment, correct?

Reply:
Yes on the sending the batch, but if that line is reached we know the previous segment has been read fully, so we are good to call the `SegmentEnd` callback. Simplifying the comment since the watcher's `SegmentEnd` callback can be used for something other.